DMA_LPC channel chaining support LPC55S69

EmilioCBen opened the pull request:
The current linking process with the LPC INPUTMUX and DMA Driver wasn't considering the parameters expected for the INPUTMUX driver. The DMA also needed a different configuration setup depending on how the DMA Chaining was used and if the user only wanted a subset of the data to be transferred.

dleach02 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -314,12 +324,31 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
-					dst_inc,
-					block_config->block_size);
+		/* If DMA Channel Chaining is expected, then the
+		 * burst length must be setup differently aswell.
```
"as well"

EmilioCBen replied:
Sorry about that, fixed now.

dleach02 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -314,12 +324,31 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		 * burst length must be setup differently aswell.
+		 */
+		if ((config->dest_chaining_en && config->source_chaining_en) ||
+		   (!config->dest_chaining_en && !config->source_chaining_en)) {
```
can we add a comment block in each of the condition blocks to describe what is going on?

EmilioCBen replied:
Thank you, I have added some additional comments, please let me know if this is more helpful.

dleach02 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -178,11 +178,21 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		INPUTMUX_AttachSignal(INPUTMUX, config->linked_channel,
+				      kINPUTMUX_Otrig0ToDma0);
+		/* Since INPUTMUX handles the signal component
+		 * to channel chaining, we should requesr that
```
"request"

EmilioCBen replied:
typos...this is fixed now.

mmahadevan108 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -314,12 +324,31 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
-					dst_inc,
-					block_config->block_size);
+		/* If DMA Channel Chaining is expected, then the
+		 * burst length must be setup differently aswell.
```
`as well`

EmilioCBen replied:
Thank you, this should be fixed now.

mmahadevan108 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -314,12 +324,31 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
-					dst_inc,
-					block_config->block_size);
+		/* If DMA Channel Chaining is expected, then the
+		 * burst length must be setup differently aswell.
```
Can you please elaborate a bit in the comment on how the burst size was picked for the different cases.

EmilioCBen replied:
I have added some additional comments, please let me know if these are more helpful.

dleach02 reviewed (dismissed):
Removing approval pending some additional discussions

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -179,12 +182,53 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
 
 	LOG_DBG("channel is %d", p_handle->channel);
 
```
you need to handle the case where the driver is re-configuring the triggering channel to not be linked to the previously linked channel anymore, for which config->linked_channel will be NULL, and at that point you should undo the inputmux connection, reconfigure the dma channel, and give the semaphore. You also need to handle the case where the currently triggering channel wants to link to a different channel, and in this case the driver should know which channel is currently using the Otrig and recognize when it is not needed to take the semaphore to reconfigure that channel

EmilioCBen replied:
I believe this case has been fixed but check just in case.

decsny replied:
@EmilioCBen still missing the case where one of the channels that is already linked to another is reconfigured to link to a different one, but there is a question of how that is supposed to work in the api imo since it isnt documented. maybe the user would want to link to two different ones, but I think for now actually let's not assume that's how the api works and only allow linking a channel to one other channel and not multiple

EmilioCBen replied:
So currently, if 
   1. A is configured to link with B
   2. A is configured to link with C
Then
   When A is finished with its transfer, both B and C fire off their transfer
Thus, 
   A can chain with multiple channels when "reconfigured", and while this behavior is _NOT_ defined in the api
   this effect _DOES_ represent what the mucx sdk sample implies when using their LPC DMA Channel Chaining Sample.
Therefore, if the user wants a different behavior, we should probably add a specific api function in the driver to unchain one of the otrig muxes.

decsny replied:
The contract for this driver implementation is defined by the Zephyr subsystem API, not what the mcuxpresso sample does.

EmilioCBen replied:
The Driver no longer allows the channel to be linked with multiple channels, instead it will "reconfigure" the channel and will remove the linkage of the channel or changed what the current channel is linked to, depending on what the user passes in.

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -44,8 +44,11 @@ struct dma_mcux_lpc_dma_data {
 	struct call_back *data_cb;
 	uint32_t *channel_index;
 	uint32_t num_channels_used;
+	char is_inputmux_setup;
```
Sure, I have made this change but as your aware and for future supporters, when we go to implement multiple chaining for the dma channel chaining, this driver will need a variable to keep track of the current OTrigs currently in use.

decsny replied:
yes and it will also need to keep track of what is connected to what to be able to disconnect things appropriately, so you would need more than this anyways

EmilioCBen replied:
This is being resolved in this PR now

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -179,12 +182,53 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			DMA_EnableChannel(DEV_BASE(dev), config->linked_channel);
+			INPUTMUX_AttachSignal(INPUTMUX, config->linked_channel,
+					kINPUTMUX_Otrig0ToDma0 );
+#endif //CONFIG_SOC_MIMXRT595_EVK
```
Good catch, this change has been made.

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -179,12 +182,53 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			DEV_BASE(dev)->CHANNEL[config->linked_channel].CFG |=
+				DMA_CHANNEL_CFG_HWTRIGEN_MASK;
+			DMA_EnableChannel(DEV_BASE(dev), config->linked_channel);
+			INPUTMUX_AttachSignal(INPUTMUX, config->linked_channel,
```
Great point,  I have fixed this!

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -179,12 +182,53 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		LOG_ERR("At the moment, this driver can only support 1 trigger"
+		"\n mux and only allows for one channel to trigger "
+		"\n other dma channels\n");
+		return -EINVAL;
```
copy/paste error, I have fixed this.

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -313,13 +357,56 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+						      src_inc,
+						      dst_inc,
+						      block_config->block_size);
+		} else if ( !k_sem_take(&is_trig_mux_in_use, K_MSEC(200))) {
```
Good point, this has been fixed.

EmilioCBen replied:
Good point, this has been fixed.

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -313,13 +357,56 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+							block_config->block_size -
+							config->source_burst_length);
+			}
+			k_sem_give(&is_trig_mux_in_use);
```
Fixed with its accompanying if statement.

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -179,12 +182,53 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
-		/* Link DMA_OTRIG 0 to channel */
-		INPUTMUX_AttachSignal(INPUTMUX, 0, config->linked_channel);
+	if ( config->source_chaining_en || config->dest_chaining_en) {
+		if ( !k_sem_take(&is_trig_mux_in_use, K_MSEC(200))) {
```
This might be a niche thing, but I prefer my approach.
In my opinion it's more readable and control flow is literally in line.
This also helps the compiler and the programmer understand the scope 
of any section of code.

decsny replied:
No, nested conditionals are much harder to read

decsny replied:
okay well at least make the timeout immediate because there is not really expected to be any reason to wait

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -179,12 +182,53 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		} else { // k_sem_take(..);
+		LOG_ERR("At the moment, this driver can only support 1 trigger"
+		"\n mux and only allows for one channel to trigger "
+		"\n other dma channels\n");
```
I prefer if my lines of code don't hit some boundary limit of characters
for at least half the screen, since this is mostly a log message, I imagine it will take 
up about the same amount of space so I added the '\n'. This prevents, in such situations (mainly chained error messages)
from having to wrap text around.

decsny replied:
this will mess up log files, there should be one log per line

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +190,43 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+#else
+			k_sem_give(&is_trig_mux_in_use);
+			LOG_ERR("\nAt this moment, this driver does not support"
+			"\n channel chaining for this SOC -> %s\n", CONFIG_SOC);
```
remove newlines, this will mess up log files, there should be one log per line

EmilioCBen replied:
Which log files mess up with the addition of a newline character?

decsny replied:
any time the output is logged

EmilioCBen replied:
While I disagree that this breaks the log message, I have taken out the new line symbols for all instances of the log messages.

decsny replied:
I cannot find another example in the whole zephyr tree of anyone putting a newline in the middle of a log message

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +190,43 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			k_sem_give(&is_trig_mux_in_use);
+		} else { /* k_sem_take(..); */
+			LOG_ERR("\nAt this moment, this driver can only support 1 trigger"
+			"\n mux and only allows for one channel to trigger "
```
remove newlines

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +190,43 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			"\n channel chaining for this SOC -> %s\n", CONFIG_SOC);
+			return -ENOTSUP;
+#endif /* CONFIG_SOC_LPC55S69_CPU0 */
+			k_sem_give(&is_trig_mux_in_use);
```
again, giving the semaphore here defeats its purpose, you need to remove this

decsny replied:
if you want to give the semaphore here, you have to change the paradigm and track whether or not a channel is using the mux in another variable rather than the semaphore being taken inherently meaning that

mmahadevan108 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +190,43 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		if (!k_sem_take(&is_trig_mux_in_use, K_MSEC(200))) {
+			LOG_DBG("link dma out 0 to channel %d", config->linked_channel);
+			/* At the moment we only have support
+			 * for OTrig and are planning to add
```
Should this be `for one OTrig`

EmilioCBen replied:
Sorry, yes this is fixed now.

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -312,13 +345,53 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
 			data->descriptor_index = data->descriptor_used;
 		}
 	} else {
```
the code for the chaining set up being in the else block of the scatter/gather implies that the two features are mutually exclusive, is this the case for LPC DMA?

EmilioCBen replied:
I don't think this comment is relevant anymore. Please correct me if I am wrong

decsny replied:
yes it looks like these two functionalities have been separated now in one of the recent pushes and are no longer mutually exclusive in the code, which is as expected

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -312,13 +345,53 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		 * burst length must be setup differently as well.
+		 */
+		if (!config->dest_chaining_en && !config->source_chaining_en) {
+			k_sem_give(&is_trig_mux_in_use);
```
giving this for any channel causes a bug where if you:
set up A to link to B
set up B
set up B to link to C
in that order, then there will be a bug where channel A loses the mux unexpectedly

mmahadevan108 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -312,13 +345,53 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+						dst_inc,
+						block_config->block_size -
+						config->source_burst_length);
+		}
```
This portion looks incorrect as we are not transferring the amount of data the user has provided via `block_size`. My suggestion would be to focus on channel chaining to trigger an event of transfer completion.

EmilioCBen replied:
this does look incorrect and should be addressed in the future of how to handle burst lengths

EmilioCBen committed:
dts: arm: nxp: lpc55S6X: Added trig bindings for DMA

Added Input/Output trigger mux address's as properties
that can be passed into the DMA driver. This is intended
to send INPUTMUX signals into the DMA.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen committed:
drivers: dma: dma_mcux_lpc: Adjusted dma driver for channel chaining.

Added to the configuration function to enable
channel chaining for the DMA_LPC that utilizes
the total of the SOC OTrig channels.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

EmilioCBen committed:
tests: drivers: dma: chan_link_transfer: Added LPC55S69 to Testcase

Added the LPC55S69 to the test case file for
dma_channel_chaining test.

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +192,72 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+					break;
+				}
+			}
+			if (!is_otrig_available) {
```
this if statement and boolean is unnecessary, you can put this code under the if on line 220 and invert the condition, then assume the rest of this block is if the trig is available since you would be returned otherwise

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +192,72 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+#else
+			k_sem_give(&is_otrig_being_configured);
+			LOG_ERR("\n\rAt this moment, this driver does not support"
+			"\n\r\tchannel chaining for this SOC -> %s\n\r", CONFIG_SOC);
```
remove newlines

EmilioCBen replied:
I have removed the newlines

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +192,72 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			k_sem_give(&is_otrig_being_configured);
+		} else { /* k_sem_take(..); */
+			LOG_ERR("\n\rDMA attempted to configure an Otrig mux"
+				"\n\r\twhile the mux was in use.\n\r");
```
remove newlines

EmilioCBen replied:
I have removed the newlines

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -491,6 +558,11 @@ static int dma_mcux_lpc_init(const struct device *dev)
 	memset(data->data_cb, 0, size_channel_data);
 
+	/* Indicate that the Otrig Muxes are not connected */
+	for (int i = 0; i < ARRAY_SIZE(data->linked_otrig_ch); ++i) {
```
why ++i? wouldnt this cause the loop to be 1-4 instead of 0-3?

EmilioCBen replied:
changed ++i to i++

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -44,8 +46,11 @@ struct dma_mcux_lpc_dma_data {
 	struct call_back *data_cb;
 	uint32_t *channel_index;
 	uint32_t num_channels_used;
+	int32_t linked_otrig_ch[4];
```
I have changed these to use 8 bit values instead.

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +192,71 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			k_sem_give(&is_otrig_being_configured);
+		} else { /* k_sem_take(..); */
+			LOG_ERR("\n\rDMA attempted to configure an Otrig mux"
+				"\n\r\twhile the mux was in use.\n\r");
```
I have removed the newlines

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +192,71 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			if (!is_otrig_available) {
+				LOG_ERR("\n\rCalling function tried to setup up otrig #%d"
+				"\n\r\twith at least two sets of channel, each otrig can"
+				"\n\r\tcan only support one pair of channels.\n\r", otrig_index);
```
I have removed the newlines

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +192,72 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+	/* Disconnecting the Otrig mux from the chaining channel */
+	if (!config->source_chaining_en && !config->dest_chaining_en) {
+		if (!k_sem_take(&is_otrig_being_configured, K_NO_WAIT)) {
+			for (int i = 0; i < ARRAY_SIZE(dma_data->linked_otrig_ch); ++i) {
```
I have changed this for consistency, but for context, it's how I'm use to writing my for loops, in the context of a for loop, there is no instruction that i is performing when it gets to the instruction i = i + 1, thus the iteration of ++i or i++ here is equivalent.

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -44,8 +46,11 @@ struct dma_mcux_lpc_dma_data {
 	struct call_back *data_cb;
 	uint32_t *channel_index;
 	uint32_t num_channels_used;
+	int32_t linked_otrig_ch[4];
```
I have made this a devicetree property now.

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +192,72 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
-		INPUTMUX_AttachSignal(INPUTMUX, 0, config->linked_channel);
+	/* Disconnecting the Otrig mux from the chaining channel */
+	if (!config->source_chaining_en && !config->dest_chaining_en) {
+		if (!k_sem_take(&is_otrig_being_configured, K_NO_WAIT)) {
```
I have added a timeout value.

hakehuang commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +194,66 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			DMA_EnableChannel(DEV_BASE(dev), config->linked_channel);
+
+			/* Link OTrig Muxes with passed-in channels */
+			INPUTMUX_AttachSignal(INPUTMUX, 0,
```
this better to be controlled by dts setting, instead of hardcode here

hakehuang commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +194,66 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
-		LOG_DBG("link dma out 0 to channel %d", config->linked_channel);
-		/* Link DMA_OTRIG 0 to channel */
-		INPUTMUX_AttachSignal(INPUTMUX, 0, config->linked_channel);
+	if (!config->source_chaining_en && !config->dest_chaining_en) {
```
usually we get channel by dma_request_channel, and in this function, we actually has a atomic bit selection for channels, and when we configure the channle actually the channel shall be filterred already. so I suggest to add a new api like dma_request_linked_channel to check whether the channel is ready to link.

hakehuang commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +194,66 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		}
+	} else {
+		if (!k_sem_take(&is_otrig_being_configured, K_MSEC(200))) {
+			LOG_DBG("link dma 0 channel %d with channel %d",
```
same as above, maybe create a new api called dma_request_link_channel()

decsny replied:
I agree that would be useful but I think an API change is outside the scope of this PR

hakehuang replied:
right

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +192,72 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			uint8_t is_otrig_available = 0;
+			uint8_t otrig_index = 0;
+
+			for (; otrig_index < ARRAY_SIZE(dma_data->linked_otrig_ch); ++otrig_index) {
```
not sure why this is outdated, it's still there

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +194,66 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			uint8_t is_otrig_available = 0;
+			uint8_t otrig_index = 0;
+
+			for (; otrig_index < dev_config->num_of_otrigs; ++otrig_index) {
```
reminding to change this because github marked the other comment as outdated and hid it

EmilioCBen replied:
Fixed.

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -484,13 +547,19 @@ static int dma_mcux_lpc_init(const struct device *dev)
 	size_channel_data =
 		sizeof(struct call_back) * config->num_of_channels;
 	data->data_cb = k_malloc(size_channel_data);
+	data->linked_otrig_ch = k_malloc(sizeof(uint8_t) * config->num_of_otrigs);
```
please don't use k_malloc, I'm doing another PR where I'm changing this driver to pre kernel and removing all the k_mallocs from the init

EmilioCBen replied:
Malloc is no longer used

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +194,66 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			 */
+			DEV_BASE(dev)->CHANNEL[config->linked_channel].CFG |=
+				DMA_CHANNEL_CFG_HWTRIGEN_MASK;
+			DMA_EnableChannel(DEV_BASE(dev), config->linked_channel);
```
instead of enabling the channel, in case it hasnt been configured by this driver, maybe you should error out if the linked channel is not set up in this driver

EmilioCBen replied:
This is the reason for the "For loop" above this statement.

decsny replied:
I don't see what you mean? What I mean is if a user calls dma_config for channel A and tries to link it to channel B, if they never called dma_config on channel B then the config call for channel A maybe should error since it can caused an unexpected behavior where a channel is triggered that isnt set up

decsny replied:
I changed my mind about this, I guess it would just add more tracking data to the driver and is already outside the contract of the API, so we can leave it to user responsibility

teburd commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -484,13 +547,19 @@ static int dma_mcux_lpc_init(const struct device *dev)
 	size_channel_data =
 		sizeof(struct call_back) * config->num_of_channels;
 	data->data_cb = k_malloc(size_channel_data);
+	data->linked_otrig_ch = k_malloc(sizeof(uint8_t) * config->num_of_otrigs);
```
k_malloc should not be used here

EmilioCBen replied:
malloc is no longer used here.

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,10 +192,71 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+				}
+			}
+			k_sem_give(&is_otrig_being_configured);
+		} else { /* k_sem_take(..); */
```
Thank you, I have cleaned this up.

decsny replied:
It looks like it didn't change but just to be clear I am not blocking over this, it is just a suggestion.

decsny commented:
@teburd can you revisit this ? the malloc is gone

dleach02 commented:
@teburd can you review the response to your change request to determine if it has been satisfactorily addressed? We are trying to get this PR cleaned up to merge before the merge window closes.

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -191,12 +202,75 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			if (dma_data->otrig_array[otrig_index].linked_channel == -1 ||
+			    dma_data->otrig_array[otrig_index].source_channel == channel) {
+				if (dma_data->otrig_array[otrig_index].source_channel == channel) {
+					DMA_DisableChannel(DEV_BASE(dev),
```
do not disable the linked channel, we are configuring this channel, not that one

EmilioCBen replied:
DMA will be persists both transfers if I do not disable the original linked channel.

decsny replied:
okay, I guess this is fine on second inspection of various files and functions

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -41,11 +44,18 @@ struct call_back {
 
+struct dma_otrig {
+	int8_t source_channel;
+	int8_t linked_channel;
```
it does not look like there is any reason to keep track of the linked channel, only the source channel, and by having this struct and extra data and lots more conditional checks involving it, the code is getting a lot more complex and hard to understand

EmilioCBen commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -191,12 +202,75 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		DEV_BASE(dev)->CHANNEL[config->linked_channel].CFG |=
+			DMA_CHANNEL_CFG_HWTRIGEN_MASK;
+
+		DMA_EnableChannel(DEV_BASE(dev), config->linked_channel);
```
If the API does not Enable the Channels, the triggers for each DMA signal will never get received and the chaining will never occur.

decsny replied:
after reviewing some code and the RM on what the enable channel means I think I understand now why this is necessary to do from here

danieldegrasse commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -191,12 +202,75 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		INPUTMUX_AttachSignal(INPUTMUX, otrig_index,
+			dev_config->otrig_base_address);
+		INPUTMUX_AttachSignal(INPUTMUX, config->linked_channel,
+				dev_config->itrig_base_address);
```
It seems this code will only work for channel 0, linked via OTRIG0.

The first call to `INPUTMUX_AttachSignal` will result in `0x0` being written to `0x160 + (otrig_index * 4)`. This will configure the OTRIG at `otrig_index` to use channel 0 as an input, as expected. *However*, if another channel is selected as the software-triggered channel to be linked, then the selected channel would still be zero, as the value written to the OTRIG inputmux register is what determines which channel is used as the input.

To resolve this, it seems you need to use following function call: `INPUTMUX_AttachSignal(INPUTMUX, otrig_index, dev_config->otrig_base_address + channel)`

The second call to `INPUTMUX_AttachSignal` will result in `0xF` being written to `0xE0 + (config->linked_channel * 4)`. This will configure channel `config->linked_channel` to use `OTRIG0` as an input. *However*, if `otrig_index != 0`, `ORIG0 ` would still be selected, which would be incorrect.

To resolve this, it seems you need to use following function call: `INPUTMUX_AttachSignal(INPUTMUX, config->linked_channel, dev_config->itrig_base_address + otrig+index);`

EmilioCBen replied:
This has been changed thank you.

danieldegrasse commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -191,12 +202,75 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+		for (otrig_index = 0; otrig_index < dev_config->num_of_otrigs; otrig_index++) {
+			if (dma_data->otrig_array[otrig_index].linked_channel != -1 &&
+			   (channel == dma_data->otrig_array[otrig_index].source_channel)) {
+				DMA_DisableChannel(DEV_BASE(dev),
```
Same comment here- I think at a minimum we need to disable the hardware trigger bit here when disabling the channel

EmilioCBen replied:
This has been changed.

danieldegrasse commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -191,12 +202,75 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			   (channel == dma_data->otrig_array[otrig_index].source_channel)) {
+				DMA_DisableChannel(DEV_BASE(dev),
+					dma_data->otrig_array[otrig_index].linked_channel);
+				dma_data->otrig_array[otrig_index].linked_channel = -1;
```
Can we use a `#define`'d constant instead of -1 here, for readability?

EmilioCBen replied:
I have added this thank you.

danieldegrasse commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -487,9 +562,21 @@ static int dma_mcux_lpc_get_status(const struct device *dev, uint32_t channel,
 	struct dma_mcux_lpc_dma_data *data = dev->data;
 	int total_dma_channels;
 
+	if (!data->data_cb || !data->otrig_array) {
```
This block needs to be removed, we do not need to check if `data_cb` and `otrig_array` are NULL as they are not allocated

EmilioCBen replied:
Removed thank you.

danieldegrasse commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -191,12 +202,75 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+				if (dma_data->otrig_array[otrig_index].source_channel == channel) {
+					DMA_DisableChannel(DEV_BASE(dev),
+						dma_data->otrig_array[otrig_index].linked_channel);
+					dma_data->num_channels_used--;
```
You cannot subtract from this variable- `num_channels_used` tracks the number of virtual channels the driver has configured. Since the driver does not offer a way to "unconfigure" a virtual channel, we cannot decrease the number of allocated channels here.

EmilioCBen replied:
These have been removed.

danieldegrasse reviewed (commented):
@EmilioCBen can you clarify the INPUTMUX_AttachSignal function calls? I want to be sure a bug does not exist here

EmilioCBen commented on dts/bindings/dma/nxp,lpc-dma.yaml:
```diff
@@ -17,6 +17,15 @@ properties:
   dma-channels:
     required: true
 
+  dma-num-of-otrigs:
```
Check this again, but I believe I made the correct change here.
